### PR TITLE
force_encoded_color flag and warning against mixed color-gray encoded DBs

### PR DESCRIPTION
**Pull request by philkr**
This PR fixes issue #1901, where older encoded DBs are not compatible with PR #1878. It adds a force_encoded_color flag to data_param, to force mixed encoded DBs to be read properly (as all color DBs).
It also warns users against the use of mixed color-gray DBs.

### CODE REVIEW

File: src/caffe/layers/data_layer.cpp [hunk 1]
```diff
@@ -98,7 +101,15 @@ void DataLayer<Dtype>::InternalThreadEntry() {
 
     cv::Mat cv_img;
     if (datum.encoded()) {
-       cv_img = DecodeDatumToCVMatNative(datum);
+      if (force_color)
+        cv_img = DecodeDatumToCVMat(datum, true);
+      else
+        cv_img = DecodeDatumToCVMatNative(datum);
+      if (cv_img.channels() != this->transformed_data_.channels())
+        LOG(WARNING) << "You dataset contains encoded images with mixed "
```
Comment (jeffdonahue) [inline src/caffe/layers/data_layer.cpp]: s/you/your/

otherwise, this looks good to me

File: src/caffe/layers/window_data_layer.cpp [hunk 2]
```diff
@@ -281,7 +281,7 @@ void WindowDataLayer<Dtype>::InternalThreadEntry() {
       if (this->cache_images_) {
         pair<std::string, Datum> image_cached =
           image_database_cache_[window[WindowDataLayer<Dtype>::IMAGE_INDEX]];
-        cv_img = DecodeDatumToCVMatNative(image_cached.second);
+        cv_img = DecodeDatumToCVMat(image_cached.second, true);
```
Comment (shelhamer) [inline src/caffe/layers/window_data_layer.cpp]: This doesn't seem right -- `ImageData` has its own color / not color flag that should be respected https://github.com/BVLC/caffe/blob/dev/src/caffe/proto/caffe.proto#L522... or maybe it should do the native decoding unless `is_color` is set to true?

**Comment by shelhamer** [comment]
@philkr thanks for working on the fix. Please remember to squash and rebase for a clean merge.

**Comment by shelhamer** [comment]
@philkr this still wasn't a clean merge, so please rebase to the latest dev next time (remember your clone might be out of date). For the sake of this being done with I did the merge manually in 9114424. Thanks.

**Comment by yanii** [comment]
Just started trying to use encoded dbs again, and after re-creating a new encoded database with the latest caffe-dev (which I see is now deprecated) the problem persists - so this doesn't seem to be a problem restricted to older databases.. even newly created DBs need the flag to force it.

Would advise testing this on a dataset with mixed image encodings - I'm using the original imagenet images (without them being re-encoded) and finding this.

**Comment by philkr** [comment]
You are right. I was trying to be a bit too smart about it in ```src/caffe/util/io.cpp:120```. I'll fix it over the weekend.

**Pull request merged**